### PR TITLE
CNV_3045 Add NIC to VM

### PR DESCRIPTION
**Pull request by bgaydosrh**
Created new module for adding NICs to VMs, conditionalized to reuse for both add to VM and VM templates.

Peer review requested from @aburdenthehand to ensure my conditionals and associated logic and syntax is valid.

Also content questions regarding GUI and associated functionality will be sent to Adam Litke and Tomas Jellinek.

Label **Peer Review Needed** and **enterprise-4.3** only.

Many thanks!

Bob

### CODE REVIEW

File: modules/cnv-add-nic-to-vm.adoc [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Module included in the following assemblies:
+//
+// * cnv/cnv_users_guide/cnv-edit-vms.adoc
+// * cnv/cnv_users_guide/cnv-editing-vm-template
+//
+// Establishing conditionals so content can be re-used for editing VMs
+// and VM templates.
+
+ifeval::["{context}" == "cnv-add-nic-to-vm"]
```
Comment (aburdenthehand) [inline modules/cnv-add-nic-to-vm.adoc]: This context needs to use the assembly id. In this case, `cnv-edit-vms`

File: modules/cnv-add-nic-to-vm.adoc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+:cnv-vm-add-nic:
+:object: virtual machine
+:object-gui: Virtual Machines
+endif::[]
+
+ifeval::["{context}" == "cnv-add-nic-to-vm-template"]
```
Comment (aburdenthehand) [inline modules/cnv-add-nic-to-vm.adoc]: And this one needs to be `cnv-editing-vm-template`

File: modules/cnv-add-nic-to-vm.adoc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+The *Link State* is set to *Up* by default when the network interface card
+is defined on the virtual machine and connected to the network.
+
+// Scrubbing all conditionals used in module
+
+:object!:
+:object-gui!:
+
+ifeval::["{context}" == "cnv-add-nic-to-vm"]
+:cnv-vm-add-nic!:
```
Comment (aburdenthehand) [inline modules/cnv-add-nic-to-vm.adoc]: Will need to change this too.

File: modules/cnv-add-nic-to-vm.adoc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+endif::[]
+
+ifeval::["{context}" == "cnv-add-nic-to-vm-template"]
```
Comment (aburdenthehand) [inline modules/cnv-add-nic-to-vm.adoc]: As above

File: modules/cnv-add-nic-to-vm.adoc (hunk 1 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done. Sorry, stupid mistake. I get it, i just forgot to make these changes after the concepts sunk in! :-)

Thanks @aburdenthehand 
Bob

File: modules/cnv-add-nic-to-vm.adoc (hunk 2 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done.

File: modules/cnv-add-nic-to-vm.adoc (hunk 3 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done.

File: modules/cnv-add-nic-to-vm.adoc (hunk 4 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done.

File: modules/cnv-add-nic-to-vm.adoc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+[id="cnv-vm-add-nic_{context}"]
+
+= Adding a network interface to a {object}
+
+.Procedure
+
+. From the *{object-gui}* tab, select the {object} you created.
+. Select the *Network Interfaces* tab.
+. Click *Add NIC*.
+. In the *Add NIC* window, specify the *Name*, *Model*, *Network*, *Type*,
```
Comment (phoracek) [inline modules/cnv-add-nic-to-vm.adoc]: It may be a question for QE, but are all those required?

Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: It appears so from the design @phoracek but best to let QE verify.
@alexxa can you verify these fields are all required in the UI, and if the UI steps are the same whether using a template or not? Thanks. Once I have that, we can merge and pick to 4.3.

Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Confirmed.

**Comment by alexxa** [comment]
@bgaydosrh there is "Add Network Interface" button, and no "Add NIC", also there is no "Restart VM" button. See the corresponding Jira for details and screenshot.

**Review by alexxa — CHANGES_REQUESTED** [review]
@bgaydosrh there is "Add Network Interface" button, and no "Add NIC", also there is no "Restart VM" button. See the corresponding Jira for details and screenshot.

File: modules/cnv-add-nic-to-vm.adoc [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+and *MAC Address* of the network interface.
+. For the virtual machine to be accessible, it must be restarted. Click *Add NIC* to add the network interface without restarting the virtual
+machine.
```
Comment (mburke5678) [inline modules/cnv-add-nic-to-vm.adoc]: This seems non-intuitive:
_it must be restarted. Click *Add NIC* to add the network interface without restarting_
The VM must be restarted but don't restart it?

File: modules/cnv-add-nic-to-vm.adoc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+. Use the drop-down lists and check boxes to edit the network
+interface configuration. When you edit an existing network interface, do
+one of the following:
+.. Click *Apply and Restart VM* to add the network interface and restart
+the virtual machine.
+.. Click *Save Changes* to add the network interface without restarting the
+virtual machine.
```
Comment (mburke5678) [inline modules/cnv-add-nic-to-vm.adoc]: Are both of these sub-steps required or are these options? 
If options, which should I pick.

**Comment by mburke5678** [comment]
A couple of questions, perhaps from my ignorance. 
I cannot comment on the metadata stuff.

**Comment by bgaydosrh** [comment]
> @bgaydosrh there is "Add Network Interface" button, and no "Add NIC", also there is no "Restart VM" button. See the corresponding Jira for details and screenshot.

"Restart VM" button. See the corresponding Jira for details and screenshot.

@alexxa - In the Jira comment, you say there is no Add NIC --- it should be Add Network Interface and the screenshots reflect that. Hoping this was a typo. See Jira record for my revision of the steps. https://issues.redhat.com/browse/CNV-3045

File: modules/cnv-add-nic-to-vm.adoc (hunk 6 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Hi @mburke5678 - This is a limitation in the current design, or seems to be. There was a button planned that auto-restarted the VM, but it's coming in another release. 

For now, I have changed the wording slightly to: Click Add Network Interface to add the network interface without restarting the virtual
machine. Note that for the virtual machine to be accessible, it must be restarted.

File: modules/cnv-add-nic-to-vm.adoc (hunk 7 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Related to your previous comment @mburke5678 --- the "Apply and Restart" button is not available in this release, so the step now reads only "Click *Save Changes*" as we have previously clarified the VM has to be restarted manually.

**Comment by bgaydosrh** [comment]
> @bgaydosrh
> A couple of questions, perhaps from my ignorance.
> I cannot comment on the metadata stuff.

Nah, not ignorance, those were both very valid comments!

**Comment by bgaydosrh** [comment]
This PR is now ready for final Peer Review and then can be merged and picked to 4.3.

A separate PR is coming for merge and pick to 4.2 as we need to rebase 4.2 files.

File: modules/cnv-add-nic-to-vm.adoc [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+= Adding a network interface to a {object}
+
+.Procedure
+
+. From the *{object-gui}* tab, select the {object} you created.
+. Select the *Network Interfaces* tab.
+. Click *Add Network Interface*.
+. In the *Add Network Interface* window, specify the *Name*, *Model*, *Network*, *Type*,
+and *MAC Address* of the network interface.
+. Click *Add Network Interface* to add the network interface without restarting the virtual
```
Comment (yossisegev) [inline modules/cnv-add-nic-to-vm.adoc]: I'm having issues in opening the web console, so I can't see or try it for myself, but this section seems weird: The user already clicked "Add Network Interface" 2 sections ago, and in the previous section he was still in the "Add Network Interface" window, so it's seems odd he should be clicking "Add Network Interface" again.
Please re-check.

File: cnv/cnv_users_guide/cnv-edit-vms.adoc [hunk 9]
```diff
@@ -11,5 +11,4 @@ a subset of the parameters in the *Virtual Machine Overview* of the web console.
 include::modules/cnv-editing-vm-web.adoc[leveloffset=+1]
 include::modules/cnv-editing-vm-yaml-web.adoc[leveloffset=+1]
 include::modules/cnv-editing-vm-cli.adoc[leveloffset=+1]
-
-include::modules/cnv-add-disk-to-vm.adoc[leveloffset=+1]
```
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-edit-vms.adoc]: @bgaydosrh I might be missing something, but did you mean to remove the add-disk module in these two assemblies?

**Review by ousleyp — COMMENTED** [review]
Same comments from PR #18801 but otherwise LGTM.

File: modules/cnv-add-nic-to-vm.adoc [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+machine. Note that for the virtual machine to be accessible, it must be restarted.
+. Use the drop-down lists and check boxes to edit the network
```
Comment (yossisegev) [inline modules/cnv-add-nic-to-vm.adoc]: 1. By "drop down lists" do you mean lists in the "Virtual Machine Details" window, or in the "Add Network Interface" sub-window.
2. No check boxes exist in neither of these 2 windows.
These errors are relevant for the 4.3 cluster web-console, which is the one I am running.

File: modules/cnv-add-nic-to-vm.adoc [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+interface configuration.
+. Click *Save Changes*.
```
Comment (yossisegev) [inline modules/cnv-add-nic-to-vm.adoc]: There is no such button (again - at least in 4.3 web console).

**Review by yossisegev — CHANGES_REQUESTED** [review]

File: modules/cnv-add-nic-to-vm.adoc (hunk 8 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Pasting in Tomas' reply as to the diff bnetween the two versions for my own reference so I can make the accorded changes for 4.3:

"on 4.3 it is:
On the nic list there is an "Add Network Interface" button to open the add nic dialog. After you fill in all the details, you need to submit the dialog; that submit button is called "Add""

Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Here is the 4.3 version:

. From the *{object-gui}* tab, select the {object}.
. Select the *Network Interfaces* tab.
. Click *Add Network Interface*.
. In the *Add Network Interface* window, specify the *Name*, *Model*, *Network*, *Type*,
and *MAC Address* of the network interface.
. Click *Add* to add the network interface.
. Restart the virtual machine to enable access.
. Edit the drop-down lists and check boxes to configure the network
interface.
. Click *Save Changes*.
. Click *OK*.

File: modules/cnv-add-nic-to-vm.adoc [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+until the user restarts it. The new network interface has a ’Pending VM restart’
+*Link State* until you restart the virtual machine. Hovering over the
+*Link State* will display more detailed information.
+
+The *Link State* is set to *Up* by default when the network interface card
+is defined on the virtual machine and connected to the network.
+
+// Scrubbing all conditionals used in module
+
+:object!:
```
Comment (aburdenthehand) [inline modules/cnv-add-nic-to-vm.adoc]: As with #18934 - This one will also need to be updated with:
ifeval::["{context}" == "cnv-edit-vms"]
:object!:
:object-gui!:
endif::[]

ifeval::["{context}" == "cnv-editing-vm-template"]
:object!:
:object-gui!:
endif::[]

File: modules/cnv-add-nic-to-vm.adoc (hunk 8 above)
Comment (yossisegev) [reply modules/cnv-add-nic-to-vm.adoc]: Here is my review after trying this walkthrough on a 4.3 cluster web-console (I don't have any comments on the section before "Click Add Network Interfaces", so I only started quoting from the section on):
> . Click _Add Network Interface_.
The button caption is _Create Network Interface_

> . In the _Add Network Interface_ window, specify the _Name_, _Model_, _Network_, _Type_,
> and _MAC Address_ of the network interface.
This one is correct - the name of the window is indeed _Add Network Interface_.

> . Click _Add_ to add the network interface.
> . Restart the virtual machine to enable access.
> . Edit the drop-down lists and check boxes to configure the network
> interface.
What drop-down lists?
I suspect this sentence refers to the drop-down lists in the _Add Network Interface_ window (Model, Network, Type). In that case - this section is both redundant and misplaced.

> . Click _Save Changes_.
> . Click _OK_.
There are no such buttons (neither _Save Changes_ nor _OK_). The additional NIC setting is applied upon exiting the _Add Network Interface_ window (by pressing _Add_ button) and restarting the VM.

Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Thanks @yossisegev, very helpful! :-)

Bob

File: cnv/cnv_users_guide/cnv-edit-vms.adoc (hunk 9 above)
Comment (bgaydosrh) [reply cnv/cnv_users_guide/cnv-edit-vms.adoc]: Yes, this PR should only have NIC but eventually will contain disk, NIC, and CDROM once all PR have been merged.

File: modules/cnv-add-nic-to-vm.adoc [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+[id="cnv-vm-add-nic_{context}"]
+
+= Adding a network interface to a {object}
+
+.Procedure
+
+. From the *{object-gui}* tab, select the {object} you created.
+. Select the *Network Interfaces* tab.
+. Click *Add Network Interface*.
```
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done now that you have clarified.

File: modules/cnv-add-nic-to-vm.adoc (hunk 10 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done now that you have clarified.

File: modules/cnv-add-nic-to-vm.adoc (hunk 11 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Done now that you have clarified.

File: modules/cnv-add-nic-to-vm.adoc (hunk 12 above)
Comment (bgaydosrh) [reply modules/cnv-add-nic-to-vm.adoc]: Added.

**Review by bgaydosrh — COMMENTED** [review]
Done.

**Pull request merged**